wlutil/build.py

```diff
@@ -186,23 +186,78 @@ def buildWorkload(cfgName, cfgs, buildBin=True, buildImg=True):
     if ret != 0:
         raise RuntimeError("Error while building workload")
 
+def makeDrivers(boardDir, linuxSrc):
+    driverDirs = pathlib.Path(boardDir).glob("drivers/*")
+    makeCmd = "make -C " + str(linuxSrc) + " ARCH=riscv CROSS_COMPILE=riscv64-unknown-linux-gnu- "
+
+    # Prepare the linux source for building external drivers
+    run(["make", "ARCH=riscv", "CROSS_COMPILE=riscv64-unknown-linux-gnu-", "modules_prepare", jlevel], cwd=linuxSrc)
+
+    drivers = []
+    for driverDir in driverDirs:
+        run(makeCmd + "M=" + str(driverDir), cwd=driverDir, shell=True)
+        drivers.extend(list(driverDir.glob("*.ko")))
+
+    return drivers
+
+def setupBoardInitramfs(boardDir, linuxSrc):
+    log = logging.getLogger()
+
+    loadScript = initramfs_root / "loadDrivers.sh"
+
+    # clear out any leftover drivers from a previous build
+    for oldDriver in initramfs_root.glob("*.ko"):
+        oldDriver.unlink()
+    if loadScript.exists():
+        loadScript.unlink()
+
+    drivers = makeDrivers(boardDir, linuxSrc)
+
+    for driverPath in drivers:
+        shutil.copy(driverPath, initramfs_root)
+
+    with open(loadScript, 'w') as f:
+        for driverPath in drivers:
+            f.write("insmod /" + str(driverPath.name) + "\n")
+
+    if not (initramfs_root / "bin" / "busybox").exists():
+        raise RunTimeError("Initramfs root not initialized. Please run './marshal init'")
+
+    # This script adds the device nodes and packages up the initrd.
+    run(['fakeroot', '--', './makeInitramfs.sh'], cwd=wlutil_dir)
+
 # Now build linux/bbl
 def makeBin(config, initramfs=False):
     log = logging.getLogger()
 
     # We assume that if you're not building linux, then the image is pre-built (e.g. during host-init)
     if 'linux-config' in config:
         linuxCfg = os.path.join(config['linux-src'], '.config')
-        shutil.copy(config['linux-config'], linuxCfg)
+        defCfg = os.path.join(config['linux-src'], 'defconfig')
+        # Create a defconfig to use as reference
+        if not os.path.isfile(defCfg):
+            run(['make', 'ARCH=riscv', 'defconfig'], cwd=config['linux-src'])
+            shutil.copy(linuxCfg, defCfg)
+            with open(defCfg, 'a') as f:
+                f.write("CONFIG_BLK_DEV_INITRD=y\n")
+                f.write('CONFIG_INITRAMFS_SOURCE="' + str(initramfs_cpio) + '"\n')
+
+        # Create a config from the user fragments
+        kconfigEnv = os.environ.copy()
+        kconfigEnv['ARCH'] = 'riscv'
+        run([os.path.join(config['linux-src'], 'scripts/kconfig/merge_config.sh'),
+            defCfg, config['linux-config']], env=kconfigEnv, cwd=config['linux-src']) 
+
+        setupBoardInitramfs(board_dir, config['linux-src'])
 
         if initramfs:
```

Comment:
This is kinda confusing in both the code and in the user experience. We used to only have an initramfs if the user wanted their rootfs in an initrd. Now, we always have an initramfs. On disk-based systems it has the driver loading stuff, on diskless systems it has your rootfs. We still use 'initramfs' in most places to mean the diskless thing, but it's not as clear as it once was.

Reply:
Anyone who's used linux on a desktop platform is probably familiar with the concept of using an initramfs to mount the real root filesystem. (https://fedoraproject.org/wiki/Dracut) or (https://wiki.debian.org/initramfs). Also along those lines, the bigger confusing/usability problem is use of the combine BBL+kernel image.

If you want to make the user experience easy, make it so unpatched Fedora & Debian kernels 'just work' in firesim. That means either porting U-boot, or refactoring the Verilator infrastructure so it can load an SBI implementation (bbl), kernel, and initramfs directly into memory and then add the addresses of the kernel and initramfs to the 'chosen' node. (see https://github.com/tmagik/freedom-u-sdk/blob/master/conf/uEnv.txt for a u-boot script example).

Reply:
I changed the terminology to be 'nodisk' now instead of 'initramfs' to make the intent more clear (the old -i --initramfs option is still there for backward compatibility). Now the term 'initramfs' anywhere in the Marshal source refers to a literal initramfs.

The confusion around BBL+kernel will need to be dealt with in the documentation. I think there's some mention of it in the docs now, but we're going to try and update stuff a bit more in anticipation of an upcoming tutorial at the Micro '19 conference.

Unpatched kernels unfortunately won't work at the moment. We are in the process of submitting patches to get the known bugs fixed, but until then, we're stuck with the patches. As much as I'd love to stop messing around with kernel sources, we're just not there yet. This PR is a big step toward allowing that though.

I'm all for getting a proper bootloader going. If someone provides a working port that plays nice with Rocket/FireChip/etc. then I'll be happy to try and integrate it. BBL clearly has limitations, but we've been able to work around them so far.